src/index.js

```diff
@@ -20,46 +18,64 @@ function response(statusCode, body) {
-    const urlencoded = event.queryStringParameters.urlencoded === "true";
+    const urlToScreenshot = event.queryStringParameters.url;
+    const delayMillis = event.queryStringParameters.delay || 4000;
+
+    const width = event.queryStringParameters.width || 800;
+    const height = event.queryStringParameters.height || 500;
+
+    const destination = event.queryStringParameters.destination;  // E.g. prod/4403/offer/1234/referrer_fulfill_20200102175601
+
+    // Given the above examples, it would end up in s3:
+    // s3://bucket/screenshots/prod/4403/offer/1234/referrer_fulfill_20200102175601.png
 
-    if (!targetUrl && !code) {
-        return response(400, "You need something to do");
+    const token = event.headers.token;
+    const SECRET_TOKEN = requireEnv.require("SECRET_TOKEN")
+
+    if (!urlToScreenshot || !destination || !token || token !== SECRET_TOKEN) {
+        return response(400, "Requires a valid url, and headers for token, destination and optionally height, width and delay");
     }
 
-    try {
-        let result = null;
-
-        switch (event.queryStringParameters.type) {
-            case "image":
-                result = await optimizeImage(targetUrl);
-                break;
-            case "code":
-                result = await screenshotCode({
-                    browser,
-                    code,
-                    codeType,
-                    urlencoded,
-                });
-                break;
-            default:
-                result = await takeScreenshot(browser, targetUrl);
-        }
+    if (!isPositiveInteger(height) || !isPositiveInteger(width) || !isPositiveInteger(delayMillis)) {
+        return response(400, "Height, width and delay must be positive integer values");
+    }
+
+    const safeUrlToScreenshot = decodeURIComponent(urlToScreenshot);
+    const safeDestinationPath = decodeURIComponent(destination);
+
+    if (!validURL(safeUrlToScreenshot)) {
+        return response(400, `url appears to be malformed (${safeUrlToScreenshot}), please specify a uri encoded string`);
+    }
 
+    const destinationFilePath = `${safeDestinationPath}.png`;
+
+    // // For keeping the browser launch
+    context.callbackWaitsForEmptyEventLoop = false;
+    const browser = await getChrome();
+
+    console.log("got browser");
+
+    try {
+        const result = await takeScreenshot(browser, safeUrlToScreenshot, destinationFilePath, delayMillis, parseInt(width), parseInt(height));
         return response(200, result);
     } catch (e) {
-        return response(500, e);
+        console.log(e)
+        return response(500, "Caught exception. Perhaps your s3 destination was invalid? Logs are available in cloudwatch");
     }
 };
+
+function validURL(str) {
+    const pattern = new RegExp('[0-9a-zA-Z_]+(\/[0-9a-zA-Z_]+)*', 'i'); // fragment locator
+    return !!pattern.test(str);
+}
+
+function isPositiveInteger(str) {
+    return parseInt(str) > 0;
```

Comment:
Contrary to popular belief, `parseInt` does NOT default to a `10` as the base. There's a gotcha around what the string stars with. So best to be explicit. Same for all the other `parseInt` uses
```suggestion
    return parseInt(str, 10) > 0;
```